jubakit/wrapper.py

```diff
@@ -0,0 +1,183 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import numpy as np
+from sklearn.base import BaseEstimator, ClassifierMixin
+from .classifier import Classifier, Config, Dataset
+
+
+class BaseJubatusClassifier(BaseEstimator, ClassifierMixin):
+  """
+  scikit-learn Wrapper for Jubatus Classifiers.
+  """
+  
+  def __init__(self, n_iter=1, shuffle=False, softmax=False, embedded=True, seed=None):
+    """
+    Creates a base class for Jubatus Classifiers"
+    """
+    self.softmax = softmax
+    self.n_iter = n_iter
+    self.shuffle = shuffle
+    self.embedded = embedded
+    self.seed = seed
+    self.fitted_ = False
+    self.clf_ = None
+
+  @classmethod
+  def _launch_classifier(self):
+    """ 
+    Launch Jubatus Classifier 
+    """
+    raise NotImplementedError()
+
+  def partial_fit(self, X, y):
+    """
+    Partially fit underlying model.
+    If underlying model does not exist, launch a new model.
+    """
+    if getattr(self, 'classes_', None) is None:
+      self.classes_ = np.unique(y)
+    if self.clf_ is None:
+      self._launch_classifier()
+      if not self.embedded:
+        self.clf_.clear()
+    dataset = Dataset.from_data(X, y)
+    for i in range(self.n_iter):
+      if self.shuffle:
+        dataset = dataset.shuffle(self.seed)
+      for _ in self.clf_.train(dataset): pass
+    self.fitted_ = True
+    return self
+
+  def fit(self, X, y):
+    """
+    Fit model.
+    """
+    self._launch_classifier()
+    if not self.embedded:
+      self.clf_.clear()
```

Comment:
- `fit` trains `X` using new classifier model.
- `partial_fit` trains `X` with existing classifier model if exists.

If these 3 lines are removed, we cannot reuse classifier instance.
For example, grid search task reuses classifier instance, but if these 3 lines are removed, classifier with new method will not launched and same cv_score will be returned.
```
clf = LinearClassifier(**args)
for method in ['AROW', 'CW']:
   clf.set_params(method=method)
   clf.fit(X_train, y_train)
   y_pred = clf.predict(X_test)
   cv_score = accuracy(y_test, y_pred)
```